### PR TITLE
Backends for S3 and local file system

### PR DESCRIPTION
**Pull request by Tyrannas**
Implementation of https://github.com/RaRe-Technologies/smart_open/issues/425

### CODE REVIEW

**Comment by gfournier** [comment]
ok maybe add a unit test by providing a mock to "sopen" function and making sure that it's called and released

File: tests/test_core.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    def __exit__(self, type, value, traceback):
+        raise FileNotFoundError
+
+def test_sopen():
+    """
+    Replace the smart_open open with a mock function to see if __enter__ and __exit__ methods are properly called.
+    """
+    backend = FileStorageBackend()
+    # mock the smart open open
+    backend.open = MagicMock(return_value=Mocker())
```
Comment (gfournier) [inline tests/test_core.py]: Use `unittest.mock.patch` function to change behavior of `smart_open.open` function.

Example with default value:
```
with unittest.mock.patch('math.sqrt', return_value=2) as _:
    assert math.sqrt(6767) == 2
```

Example with special function (can be used to set a market for example):
```
mock_called = False
def myMock(x):
    global mock_called
    mock_called = True
    return x**2

with unittest.mock.patch('math.sqrt', new=myMock) as _:
    assert math.sqrt(6767) == 6767**2
    assert mock_called
```

**Review by gfournier — CHANGES_REQUESTED** [review]

File: tests/test_core.py (hunk 1 above)
Comment (Tyrannas) [reply tests/test_core.py]: https://docs.python.org/3/library/unittest.mock.html#quick-guide I thought MagiMock would do this ?

**Comment by gfournier** [comment]
Ok for the tests.

One more value to test: can we make sure that `smart_open` is called with the right parameters ?
For example `s3://` URL for S3 backend, `file://` URL for filesystem backend, etc...

It will test that the right URL is called for each backend.

+ same for connexion parameters

**Review by gfournier — APPROVED** [review]

**Pull request merged**